service/history/queues/slice.go

```diff
@@ -0,0 +1,370 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queues
+
+import (
+	"fmt"
+
+	ctasks "go.temporal.io/server/common/tasks"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+type (
+	Slice interface {
+		Scope() Scope
+		CanSplitByRange(tasks.Key) bool
+		SplitByRange(tasks.Key) (left Slice, right Slice)
+		SplitByPredicate(tasks.Predicate) (pass Slice, fail Slice)
+		CanMergeByRange(Slice) bool
+		MergeByRange(Slice) Slice
+		CanMergeByPredicate(Slice) bool
+		MergeByPredicate(Slice) Slice
+		ShrinkRange()
+		SelectTasks(int) ([]Executable, error)
+	}
+
+	executableInitializer func(tasks.Task) Executable
+
+	SliceImpl struct {
+		executableInitializer executableInitializer
+
+		destroyed bool
+
+		scope     Scope
+		iterators []Iterator
+
+		// TODO: make task tracking a separate component
+		// and evaluate the performance of using a btree
+		// for storing executables
+		outstandingExecutables map[tasks.Key]Executable
+	}
+)
+
+func NewSlice(
+	paginationFnProvider paginationFnProvider,
+	executableInitializer executableInitializer,
+	scope Scope,
+) *SliceImpl {
+	return &SliceImpl{
+		executableInitializer:  executableInitializer,
+		scope:                  scope,
+		outstandingExecutables: make(map[tasks.Key]Executable),
+		iterators: []Iterator{
+			NewIterator(paginationFnProvider, scope.Range),
+		},
+		destroyed: false,
+	}
+}
+
+func (s *SliceImpl) Scope() Scope {
+	s.validateNotDestroyed()
+	return s.scope
+}
+
+func (s *SliceImpl) CanSplitByRange(key tasks.Key) bool {
+	s.validateNotDestroyed()
+	return s.scope.CanSplitByRange(key)
+}
+
+func (s *SliceImpl) SplitByRange(key tasks.Key) (leftSlice Slice, rightSlice Slice) {
+	if !s.CanSplitByRange(key) {
+		panic(fmt.Sprintf("Unable to split queue slice with range %v at %v", s.scope.Range, key))
+	}
+
+	leftScope, rightScope := s.scope.SplitByRange(key)
+	leftExecutables, rightExecutables := s.splitExecutables(leftScope, rightScope)
+
+	leftIterators := make([]Iterator, 0, len(s.iterators)/2)
+	rightIterators := make([]Iterator, 0, len(s.iterators)/2)
```

Comment:
if len is odd number, would this result in short of one slot?

Reply:
ok, it does not matter that much, the append will create new slice if it is not enough.
But it is almost certain that at least one of them will be too short and will have to create new slice and copy over, but maybe that is fine.

Reply:
yeah, it's just an estimation. The exact number doesn't matter too much.